File: dkplayer-ui/src/main/java/xyz/doikki/videocontroller/component/LiveControlView.java
package xyz.doikki.videocontroller.component;

import android.app.Activity;
import android.content.Context;
import android.content.pm.ActivityInfo;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.animation.Animation;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import xyz.doikki.videocontroller.R;
import xyz.doikki.videoplayer.DKVideoView;
import xyz.doikki.videoplayer.controller.component.ControlComponent;
import xyz.doikki.videoplayer.util.PlayerUtils;

/**
 * 直播底部控制栏
 * <p>
 * 此控件不适配TV模式
 */
public class LiveControlView extends BaseControlComponent implements ControlComponent, View.OnClickListener {

    private ImageView mFullScreen;
    private LinearLayout mBottomContainer;
    private ImageView mPlayButton;

    public LiveControlView(@NonNull Context context) {
        super(context);
    }

    public LiveControlView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public LiveControlView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void setupViews() {
        setVisibility(GONE);
        LayoutInflater.from(getContext()).inflate(R.layout.dkplayer_layout_live_control_view, this, true);
        mFullScreen = findViewById(R.id.fullscreen);
        mFullScreen.setOnClickListener(this);
        mBottomContainer = findViewById(R.id.bottom_container);
        mPlayButton = findViewById(R.id.iv_play);
        mPlayButton.setOnClickListener(this);
        ImageView refresh = findViewById(R.id.iv_refresh);
        refresh.setOnClickListener(this);
    }

    @Override
    public void onVisibilityChanged(boolean isVisible, Animation anim) {
        if (isVisible) {
            if (getVisibility() == GONE) {
                setVisibility(VISIBLE);
                if (anim != null) {
                    startAnimation(anim);
                }
            }
        } else {
            if (getVisibility() == VISIBLE) {
                setVisibility(GONE);
                if (anim != null) {
                    startAnimation(anim);
                }
            }
        }
    }

    @Override
    public void onPlayStateChanged(int playState) {
        switch (playState) {
            case DKVideoView.STATE_IDLE:
            case DKVideoView.STATE_START_ABORT:
            case DKVideoView.STATE_PREPARING:
            case DKVideoView.STATE_PREPARED:
            case DKVideoView.STATE_ERROR:
            case DKVideoView.STATE_PLAYBACK_COMPLETED:
                setVisibility(GONE);
                break;
            case DKVideoView.STATE_PLAYING:
                mPlayButton.setSelected(true);
                break;
            case DKVideoView.STATE_PAUSED:
                mPlayButton.setSelected(false);
                break;
            case DKVideoView.STATE_BUFFERING:
            case DKVideoView.STATE_BUFFERED:
                mPlayButton.setSelected(mControlWrapper.isPlaying());
                break;
        }
    }

    @Override
    public void onScreenModeChanged(int screenMode) {
        switch (screenMode) {
            case DKVideoView.SCREEN_MODE_NORMAL:
                mFullScreen.setSelected(false);
                break;
            case DKVideoView.SCREEN_MODE_FULL:
                mFullScreen.setSelected(true);
                break;
        }

        Activity activity = PlayerUtils.scanForActivity(getContext());
        if (activity != null && mControlWrapper.hasCutout()) {
            int orientation = activity.getRequestedOrientation();
            int cutoutHeight = mControlWrapper.getCutoutHeight();
            if (orientation == ActivityInfo.SCREEN_ORIENTATION_PORTRAIT) {
                mBottomContainer.setPadding(0, 0, 0, 0);
            } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE) {
                mBottomContainer.setPadding(cutoutHeight, 0, 0, 0);
            } else if (orientation == ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE) {
                mBottomContainer.setPadding(0, 0, cutoutHeight, 0);
            }
        }
    }

    @Override
    public void onLockStateChanged(boolean isLocked) {
        onVisibilityChanged(!isLocked, null);
    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.fullscreen) {
            mControlWrapper.toggleFullScreen();
        } else if (id == R.id.iv_play) {
            mControlWrapper.togglePlay();
        } else if (id == R.id.iv_refresh) {
            mControlWrapper.replay(true);
        }
    }
}

File: dkplayer-ui/src/main/java/xyz/doikki/videocontroller/component/LiveControlView.kt
package xyz.doikki.videocontroller.component

import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.ActivityInfo
import android.util.AttributeSet
import android.view.View
import android.view.animation.Animation
import android.widget.ImageView
import android.widget.LinearLayout
import xyz.doikki.videocontroller.R
import xyz.doikki.videoplayer.VideoView
import xyz.doikki.videoplayer.util.PlayerUtils
import xyz.doikki.videoplayer.util.orDefault

/**
 * 直播底部控制栏
 *
 *
 * 此控件不适配TV模式
 */
class LiveControlView @JvmOverloads constructor(
    context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0
) : BaseControlComponent(context, attrs, defStyleAttr), View.OnClickListener {

    private val fullScreen: ImageView
    private val bottomContainer: LinearLayout
    private val playButton: ImageView

    override fun onVisibilityChanged(isVisible: Boolean, anim: Animation?) {
        if (isVisible) {
            if (visibility == GONE) {
                visibility = VISIBLE
                anim?.let { startAnimation(it) }
            }
        } else {
            if (visibility == VISIBLE) {
                visibility = GONE
                anim?.let { startAnimation(it) }
            }
        }
    }

    override fun onPlayStateChanged(playState: Int, extras: HashMap<String, Any>) {
        when (playState) {
            VideoView.STATE_IDLE, VideoView.STATE_START_ABORT, VideoView.STATE_PREPARING, VideoView.STATE_PREPARED, VideoView.STATE_ERROR, VideoView.STATE_PLAYBACK_COMPLETED -> visibility =
                GONE
            VideoView.STATE_PLAYING -> playButton.isSelected = true
            VideoView.STATE_PAUSED -> playButton.isSelected = false
            VideoView.STATE_BUFFERING, VideoView.STATE_BUFFERED -> playButton.isSelected =
                controller?.playerControl?.isPlaying.orDefault()
        }
    }

    @SuppressLint("SwitchIntDef")
    override fun onScreenModeChanged(screenMode: Int) {
        when (screenMode) {
            VideoView.SCREEN_MODE_NORMAL -> fullScreen.isSelected = false
            VideoView.SCREEN_MODE_FULL -> fullScreen.isSelected = true
        }
        val activity = PlayerUtils.scanForActivity(context)
        val controller = controller
        if (activity != null && controller != null && controller.hasCutout == true) {
            val orientation = activity.requestedOrientation
            val cutoutHeight = controller.cutoutHeight
            when (orientation) {
                ActivityInfo.SCREEN_ORIENTATION_PORTRAIT -> {
                    bottomContainer.setPadding(0, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE -> {
                    bottomContainer.setPadding(cutoutHeight, 0, 0, 0)
                }
                ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE -> {
                    bottomContainer.setPadding(0, 0, cutoutHeight, 0)
                }
            }
        }
    }

    override fun onLockStateChanged(isLocked: Boolean) {
        onVisibilityChanged(!isLocked, null)
    }

    override fun onClick(v: View) {
        controller?.let { controller ->
            when (v.id) {
                R.id.fullscreen -> {
                    controller.toggleFullScreen()
                }
                R.id.iv_play -> {
                    controller.togglePlay()
                }
                R.id.iv_refresh -> {
                    controller.replay(true)
                }
                else -> {}
            }
        }

    }

    init {
        visibility = GONE
        layoutInflater.inflate(R.layout.dkplayer_layout_live_control_view, this)
        fullScreen = findViewById(R.id.fullscreen)
        fullScreen.setOnClickListener(this)
        bottomContainer = findViewById(R.id.bottom_container)
        playButton = findViewById(R.id.iv_play)
        playButton.setOnClickListener(this)
        val refresh = findViewById<ImageView>(R.id.iv_refresh)
        refresh.setOnClickListener(this)
    }
}